app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -33,33 +34,51 @@ def data_for_breadcrumbs
           breadcrumbs.push(:title => @title)
         end
       else
+        self.x_node_text = params[:text] if action_name == "tree_select" # get text of the active node
+
         # Append breadcrumb from title of the accordion (eg "Policies")
-        breadcrumbs.push(:title => accord_title, :key => "#{accord_name}_accord", :action => "accordion_select") if accord_title
+        breadcrumbs.push(:title => accord_title, :key => "root") if accord_title
+
+        # On special page (tagitems, politems, etc.)
+        if @tagitems || @politems || @ownershipitems || @retireitems
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems, true)) unless options[:hide_special_item]
+          breadcrumbs.push(:title => @right_cell_text)
+        else
+          # Ancestry parents breadcrumbs (only in services)
+          breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
 
-        # Append breadcrumbs created from the tree (eg "All policies > Red Hat policies > Policy 1")
-        breadcrumbs_from_tree = build_breadcrumbs_from_tree
-        breadcrumbs.concat(breadcrumbs_from_tree)
+          # Try to create treenode from the x_node
+          entity, entity_id, _ = TreeBuilder.extract_node_model_and_id(x_node) if x_node
+          record = entity.try(:constantize).try(:find, entity_id)
+          title = TreeNode.new(record).text if record
 
-        if (options[:include_record] || breadcrumbs_from_tree.blank?) && options[:record_info].present? && options[:record_info][options[:record_title]] &&
-           !(@tagitems || @politems || @ownershipitems || @retireitems)
-          # Append breadcrumb created from record_info if included or if result from tree was empty (eg filters page)
+          # otherwise get the text from different sources
+          unless title
```

Comment:
I think we should reorganize the conditionals below to not do extra work when not necessary. I mean, with this code below there's a chance that we're setting the `title` multiple times. Would it be possible to prevent this from happening?

Reply:
I reorganized the whole sequence according to priorities of ways how to get the node text.